src/tool/xmeta/xmeta_emit.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+#include "xmeta_models.h"
+#include "xlang_model_listener.h"
+#include <winrt/base.h>
+#include <vector>
+#include <mscoree.h>
+#include <cor.h>
+#include <meta_reader.h>
+#include <meta_writer.h>
+
+namespace xlang::xmeta
+{
+    inline std::string type_semantics_to_string(model_ref<type_semantics> const& semantic_type);
+
+    class xmeta_emit : public xlang_model_listener
+    {
+    public:
+        xmeta_emit(std::string const& assembly_name) 
+            : m_assembly_name(assembly_name)
+        { };
+
+        void initialize();
+        void uninitialize();
+        void save_to_file() const;
+		std::vector<uint8_t> save_to_memory() const;
+
+        void listen_namespace_model(std::shared_ptr<namespace_model> const& model) override;
+        void listen_class_model(std::shared_ptr<class_model> const& model) override;
+        void listen_struct_model(std::shared_ptr<struct_model> const& model) override;
+        void listen_interface_model(std::shared_ptr<interface_model> const& model) override;
+        void listen_enum_model(std::shared_ptr<enum_model> const& model) override;
+        void listen_delegate_model(delegate_model const& model) override;
+
+        mdTypeDef define_type_def(std::string const& name, DWORD const& type_flag, mdToken token_extend, mdToken token_implements[]);
```

Comment:
Should these "define" methods be public? They seem more appropriate as private methods.

Reply:
Yea I think we can make them private. There is no situation here where we have to call these define methods from the walker. Unless the walker is also walking through methods, properties, and events.